cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
+            # we need to transpose original array so all index arrays
+            # are close to each other
+            transpose_needed = False
+            transpose_indices = tuple()
+            # since we can't call Copy operation on transformed Store, after
+            # the transformation, we need to return a copy
             tuple_of_arrays = ()
-            for k in key:
-                if not isinstance(k, NumPyThunk):
-                    raise NotImplementedError(
-                        "need support for mixed advanced indexing"
+            index_map = []
+
+            # First, we need to check if transpose is needed
+            for dim, k in enumerate(key):
+                if np.isscalar(k) or isinstance(k, NumPyThunk):
+                    if start_index == -1:
+                        start_index = dim
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
+            else:
+                index_map = tuple(range(len(key)))
+
+            for d, k in enumerate(key):
+                dim = index_map[d]
+                if np.isscalar(k):
+                    if k < 0:
+                        k += store.shape[dim + shift]
+                    store = store.project(dim + shift, k)
+                    shift -= 1
+                elif k is np.newaxis:
+                    store = store.promote(dim + shift, 1)
+                elif isinstance(k, slice):
+                    store = store.slice(dim + shift, k)
+                elif isinstance(k, NumPyThunk):
+                    if k.dtype == np.bool:
+                        if k.shape[0] != store.shape[dim]:
+                            raise ValueError(
+                                "boolean index did not match "
+                                "indexed array along dimension  "
+                            )
+                        # in case of the mixed indises we all nonzero
+                        # for the bool array
+                        k = k.nonzero()
+                        tuple_of_arrays += k
+                    else:
+                        tuple_of_arrays += (self.runtime.to_deferred_array(k),)
+                else:
+                    raise TypeError(
+                        "Unsupported entry type passed to advanced",
```

Comment:
Leave a space at the end of this string, to concatenate properly with the next one.